src/userSurvey.js

```diff
@@ -190,37 +295,106 @@ export async function setSurvey(options) {
         await storageSpace.set("lastSurveyRequest", lastSurveyRequest);
         await storageSpace.set("surveyCompleted", false);
         await storageSpace.set("surveyCancelled", false);
-        await storageSpace.set("surveyId", id.generateId());
         openSurveyInNewTab();
     }
 
     // Schedule a reminder for the user
     scheduleReminderForUser();
 
-    // Set a listener for the survey completion URL
+    // Set a listener for the survey completion URL.
     browser.webRequest.onBeforeRequest.addListener(
-        () => {
-            storageSpace.set("surveyCompleted", true);
-            setPopupToNoPromptPage();
-        },
-        { urls: [ (new URL(options.surveyCompletionUrl)).href + "*" ] }
+        completeSurvey,
+        { urls: [ (new URL(surveyDetails.surveyCompletionUrl)).href + "*" ] }
     );
 
-    // Set listeners for cancel and open survey button clicks in the survey request
-    messaging.onMessage.addListener(() => {
-        storageSpace.set("surveyCancelled", true);
-        setPopupToNoPromptPage();
-    }, { type: "webScience.userSurvey.cancelSurvey" });
-    messaging.onMessage.addListener(openSurveyInNewTab, { type: "webScience.userSurvey.openSurvey" });
+    // Listeners for cancel and open survey button click only need to be added once.
+    // They do not need to be added again for subsequent calls to setSurvey.
+    // These listeners do not need to be removed in endCurrentSurvey because they will
+    // not receive messages when the popup is the no prompt page.
+    if (!listenersRegistered) {
+        // Set listeners for cancel and open survey button clicks in the survey request.
+        messaging.onMessage.addListener(() => {
+            storageSpace.set("surveyCancelled", true);
+            setPopupToNoPromptPage();
```

Comment:
Need to remove the survey completion URL listener if the user canceled the survey.